### ISSUE TITLE
Emit the x-axis range when one of the Generation spinner buttons is pressed.  

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/natural-selection#43]
From https://github.com/phetsims/natural-selection/issues/34#issuecomment-567680890...

While the sim is playing, the Population graph's x-axis range autoscrolls on each tick of the clock, so `xRangeProperty` is `phetioHighFrequency: true`.  But we want to see the x-axis range in the data stream when the sim is paused and the user is controlling the range. 

Emit the x-axis range when one of the Generation spinner buttons is pressed.

**Comment by pixelzoom** [comment]
In the above commit, I added an Action to GenerationScrollControl.   This control appears below the Population graph, and control the range of the x axis:

<img width="611" alt="screenshot_50" src="https://user-images.githubusercontent.com/3046552/71864498-2a9b6280-30bd-11ea-8712-6a0425ea4dd7.png">

Here's the Action:

```js
      // Sends a PhET-iO message when the generation range is changed by the user using one of the arrow buttons.
      // See https://github.com/phetsims/natural-selection/issues/43
      const changeGenerationRangeAction = new Action( function( min, max ) {
        rangeProperty.value.setMinMax( min, max ); // mutate value
        rangeProperty.notifyListenersStatic(); // force notification
      }, {
        parameters: [
          { name: 'min', phetioType: NumberIO },
          { name: 'max', phetioType: NumberIO }
        ],
        tandem: options.tandem.createTandem( 'changeGenerationRangeAction' ),
        phetioDocumentation: 'When the user changes the generation range, this emits the new range.',
        phetioReadOnly: true,
        phetioEventType: EventType.USER
      } );
```

The Action's `execute` method has 2 NumberIO params instead of 1 RangeIO param because we are mutating the value of `rangeProperty` -- it's high frequency, and allocating Range instances would be an undesirable cost. 

Here's what I see in the console with the Data:colorized wrapper:

```
naturalSelection.labScreen.view.populationNode.populationGraphNode.generationScrollControl.changeGenerationRangeAction emitted {"min":6,"max":11}
```

It looks like this in Studio:

<img width="701" alt="screenshot_49" src="https://user-images.githubusercontent.com/3046552/71863716-03439600-30bb-11ea-8f65-63c318b18b8a.png">

Before I request design review, I'd like a sanity check from someone on the PhET-iO sub team.  

* Is this the pattern that you recommend?
* Is `phetioReadOnly: true` appropriate?
* Is it proper to say "emit" in this context (e.g. in `phetioDocumentation`)? There is no Emitter involved here. Or is there a better way  of saying "put a message in the data stream"?
* Any other feedback?

**Comment by chrisklus** [comment]
In the above commit we fixed a bug in Property where `notifyListenersStatic` wasn't emitting to the data stream like `_notifyListeners` does, so we changed its implementation. Now that that's fixed, the `xRangeProperty` is emitting to the data stream when it's mutated, so it would be superfluous to have an additional Action to emit the same range values.

The following patch removes the Action:

<details>

```diff
Index: js/phet-io/natural-selection-phet-io-elements-baseline.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/phet-io/natural-selection-phet-io-elements-baseline.js	(revision 8b01dacaac9a89e67f891b316db267efd0b9b08e)
+++ js/phet-io/natural-selection-phet-io-elements-baseline.js	(date 1578511113000)
@@ -6332,19 +6332,6 @@
       "phetioStudioControl": true,
       "phetioTypeName": "PropertyIO<BooleanIO>"
     },
-    "naturalSelection.labScreen.view.populationNode.populationGraphNode.generationScrollControl.changeGenerationRangeAction": {
-      "phetioDocumentation": "When the user changes the generation range, this emits the new range. The arguments are:<br/><ol><li>min: NumberIO</li><br/><li>max: NumberIO</li></ol>",
-      "phetioDynamicElement": false,
-      "phetioEventType": "USER",
-      "phetioFeatured": false,
-      "phetioHighFrequency": false,
-      "phetioIsArchetype": false,
-      "phetioPlayback": false,
-      "phetioReadOnly": true,
-      "phetioState": false,
-      "phetioStudioControl": true,
-      "phetioTypeName": "ActionIO<NumberIO, NumberIO>"
-    },
     "naturalSelection.labScreen.view.populationNode.populationGraphNode.generationScrollControl.forwardButton": {
       "phetioDocumentation": "",
       "phetioDynamicElement": false,
Index: js/common/view/population/GenerationScrollControl.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/view/population/GenerationScrollControl.js	(revision 8b01dacaac9a89e67f891b316db267efd0b9b08e)
+++ js/common/view/population/GenerationScrollControl.js	(date 1578511607000)
@@ -9,14 +9,11 @@
   'use strict';
 
   // modules
-  const Action = require( 'AXON/Action' );
   const ArrowButton = require( 'SUN/buttons/ArrowButton' );
-  const EventType = require( 'TANDEM/EventType' );
   const merge = require( 'PHET_CORE/merge' );
   const HBox = require( 'SCENERY/nodes/HBox' );
   const naturalSelection = require( 'NATURAL_SELECTION/naturalSelection' );
   const NaturalSelectionConstants = require( 'NATURAL_SELECTION/common/NaturalSelectionConstants' );
-  const NumberIO = require( 'TANDEM/types/NumberIO' );
   const Property = require( 'AXON/Property' );
   const Tandem = require( 'TANDEM/Tandem' );
   const Text = require( 'SCENERY/nodes/Text' );
@@ -53,28 +50,18 @@
         maxWidth: 120 // determined empirically
       } );
 
-      // Sends a PhET-iO message when the generation range is changed by the user using one of the arrow buttons.
-      // See https://github.com/phetsims/natural-selection/issues/43
-      const changeGenerationRangeAction = new Action( function( min, max ) {
+      // mutate the range and notify listeners
+      const updateRange = max => {
+        const min = max - rangeLength;
         rangeProperty.value.setMinMax( min, max ); // mutate value
         rangeProperty.notifyListenersStatic(); // force notification
-      }, {
-        parameters: [
-          { name: 'min', phetioType: NumberIO },
-          { name: 'max', phetioType: NumberIO }
-        ],
-        tandem: options.tandem.createTandem( 'changeGenerationRangeAction' ),
-        phetioDocumentation: 'When the user changes the generation range, this emits the new range.',
-        phetioReadOnly: true,
-        phetioEventType: EventType.USER
-      } );
+      };
 
       // back button
       const back = () => {
         isPlayingProperty.value = false; // pause the sim when we scroll back
         const max = Math.ceil( rangeProperty.value.max - options.step ); // snap to integer value
-        const min = max - rangeLength;
-        changeGenerationRangeAction.execute( min, max );
+        updateRange( max );
       };
       const backButton = new ArrowButton( 'left', back,
         merge( {
@@ -85,8 +72,7 @@
       // forward button
       const forward = () => {
         const max = Math.floor( rangeProperty.value.max + options.step ); // snap to integer value
-        const min = max - rangeLength;
-        changeGenerationRangeAction.execute( min, max );
+        updateRange( max );
       };
       const forwardButton = new ArrowButton( 'right', forward,
         merge( {
@@ -104,7 +90,7 @@
         [ rangeProperty, maxProperty ],
         ( range, max ) => {
           backButton.enabled = ( range.min > 0 );
-          forwardButton.enabled = ( range.max < max  );
+          forwardButton.enabled = ( range.max < max );
         } );
 
       // @private

```

</details>

```json
{
  "index": 6827,
  "time": 1578511261149,
  "type": "USER",
  "phetioID": "naturalSelection.general.controller.input.mouseUpAction",
  "name": "emitted",
  "componentType": "ActionIO<Vector2IO, EventIO>",
  "data": {
    "point": {
      "x": 488.61328125,
      "y": 720.75390625
    },
    "event": {
      "pointerId": 1,
      "pointerType": "mouse",
      "clientX": 488.61328125,
      "clientY": 720.75390625,
      "button": 0,
      "which": 1,
      "target": {}
    }
  },
  "metadata": {
    "dataKeys": [
      "point",
      "event"
    ],
    "playback": true
  },
  "children": [
    {
      "index": 6828,
      "time": 1578511261149,
      "type": "USER",
      "phetioID": "naturalSelection.labScreen.view.populationNode.populationGraphNode.generationScrollControl.backButton.pressListener.releaseAction",
      "name": "emitted",
      "componentType": "ActionIO<NullableIO<SceneryEventIO>>",
      "data": {
        "event": {
          "type": "up",
          "domEventType": "pointerup",
          "point": {
            "x": 488.61328125,
            "y": 720.75390625
          }
        }
      },
      "children": [
        {
          "index": 6829,
          "time": 1578511261149,
          "type": "USER",
          "phetioID": "naturalSelection.labScreen.view.populationNode.populationGraphNode.generationScrollControl.backButton.firedEmitter",
          "name": "emitted",
          "componentType": "EmitterIO<>",
          "children": [
            {
              "index": 6830,
              "time": 1578511261149,
              "type": "MODEL",
              "phetioID": "naturalSelection.labScreen.model.populationModel.xRangeProperty",
              "name": "changed",
              "componentType": "PropertyIO<RangeIO>",
              "data": {
                "oldValue": null,
                "newValue": {
                  "min": 3,
                  "max": 8
                }
              }
            }
          ]
        }
      ]
    }
  ]
}
```

Note that there is no `oldValue` in the data stream after our fix in the commit above. This is because of the mutation pattern being used with `notifyListenersStatic`. We wanted to ask if you considered using some sort of pool instead, so that `oldValue` is in the data stream and any `link` sites. We identified the possible ways to use Range for this case:

* create `new Range` every frame. We did a quick test by creating 1000 Ranges per frame, and found that the garbage collection frequency changed from ~1s to ~700 ms. The creation of Ranges doesn't show up on the Profiler until we increased the number created per frame to 10000.
* create a "manual pool" with 2 Ranges
* modify Range.js to mixin with Poolable, and then use `createFromPool` and `freeToPool`
* the current mutation pattern

In summary, the phet-io team recommends:
1. removing the `changeGenerationRangeAction` Action.
2. We think it'd be good to either not use a mutable pattern or clearly document that `xRangeProperty` does not emit `oldValue` (including in `phetioDocumentation` so that phet-io clients know not to listen to `oldValue` in `link` and `lazyLink`).

**Comment by pixelzoom** [comment]
> Now that that's fixed, the `xRangeProperty` is emitting to the data stream when it's mutated, so it would be superfluous to have an additional Action to emit the same range values.

Let me clarify...  `xRangeProperty` maybe be updated either automatically (on every clock tick) as the Population graph scrolls.  Or it may be updated manually by the user, using the Generations spinners below the Population graph.  Because it may update on every clock tick, `xRangeProperty` is `phetioHighFrequency: true`.   The designers would like manual control of `xRangeProperty` (via the spinners) to result in the new range appearing in the data stream.  Hence the need for a Action.

**Comment by pixelzoom** [comment]
I discussed the recommendations in https://github.com/phetsims/natural-selection/issues/43#issuecomment-572233507 with @samreid @zepumph and @chrisklus via Slack.

> 1. removing the `changeGenerationRangeAction` Action.

The new PhET-iO feature to be implemented in https://github.com/phetsims/phet-io/issues/1603 makes the Action unnecessary.   In a nutshell, `xRangeProperty` (or any other  `phetioHighFrequency: true`) will still emit a message if it's nested under something that is `phetioHighFrequency: false`.   So when the x-axis is automatically scrolling, messages will be filtered out; when the user changes the range via the spinners, a message will appear in the data stream. In the above commit, I've deleted the `changeGenerationRangeAction` Action.

> 2. We think it'd be good to either not use a mutable pattern or clearly document that `xRangeProperty` does not emit `oldValue` (including in `phetioDocumentation` so that phet-io clients know not to listen to `oldValue` in `link` and `lazyLink`).

I agree that mutating the `xRangeProperty` value is to be avoided, I haven't liked that solution from the start.  Having to manage my own pool, or add pooling to `Range`, I like even less.  So I'm going to go with creating new `Range` instances, and see what happens to performance when data is added to the graph (change made in above commit).   If there's a performance problem, then I'll circle back and try to convince the design team that this should be `{Property.<number>} xMaximumProperty`, since the x-axis width is constant and xMin can be computed.

**Comment by pixelzoom** [comment]
This issue now requires completion of https://github.com/phetsims/phet-io/issues/1603.  I'll leave open and unassigned for now.

**Comment by pixelzoom** [comment]
phetsims/phet-io#1603 has been addressed and closed.  I tested in Natural Selection, and it's working nicely for `xRangeProperty`.  Closing.
